File: synth_ai/sdk/task/datasets.py
"""Dataset registry and helpers shared by Task Apps."""

from __future__ import annotations

from collections.abc import Callable, Hashable
from typing import Any, List

from pydantic import BaseModel, Field, field_validator


class TaskDatasetSpec(BaseModel):
    """Declarative metadata describing a dataset that a Task App exposes."""

    id: str
    name: str
    version: str | None = None
    splits: list[str] = Field(default_factory=list)
    default_split: str | None = None
    cardinality: int | None = None
    description: str | None = None

    @field_validator("default_split")
    @classmethod
    def _validate_default_split(cls, value: str | None, info):
        values = info.data if hasattr(info, "data") else {}  # type: ignore[attr-defined]
        if value and value not in (values.get("splits") or []):
            raise ValueError("default_split must be one of splits when provided")
        return value


RegistryLoader = Callable[[TaskDatasetSpec], Any]


class TaskDatasetRegistry:
    """Lightweight registry mapping dataset specs to loader callables."""

    def __init__(self) -> None:
        self._entries: dict[str, tuple[TaskDatasetSpec, RegistryLoader, bool]] = {}
        self._cache: dict[Hashable, Any] = {}

    def register(
        self, spec: TaskDatasetSpec, loader: RegistryLoader, *, cache: bool = True
    ) -> None:
        """Register a dataset loader and its metadata."""

        self._entries[spec.id] = (spec, loader, cache)

    def describe(self, dataset_id: str) -> TaskDatasetSpec:
        if dataset_id not in self._entries:
            raise KeyError(f"Dataset not registered: {dataset_id}")
        spec, _, _ = self._entries[dataset_id]
        return spec

    def list(self) -> List[TaskDatasetSpec]:
        return [entry[0] for entry in self._entries.values()]

    def get(self, spec: TaskDatasetSpec | str) -> Any:
        """Return dataset materialisation (with optional caching)."""

        if isinstance(spec, str):
            if spec not in self._entries:
                raise KeyError(f"Dataset not registered: {spec}")
            base_spec, loader, cache_enabled = self._entries[spec]
            effective_spec = base_spec
        else:
            if spec.id not in self._entries:
                raise KeyError(f"Dataset not registered: {spec.id}")
            base_spec, loader, cache_enabled = self._entries[spec.id]
            effective_spec = base_spec.model_copy(update=spec.model_dump(exclude_unset=True))

        cache_key: Hashable = (
            effective_spec.id,
            effective_spec.version,
            effective_spec.default_split,
        )
        if cache_enabled:
            if cache_key not in self._cache:
                self._cache[cache_key] = loader(effective_spec)
            return self._cache[cache_key]
        return loader(effective_spec)

    @staticmethod
    def ensure_split(spec: TaskDatasetSpec, split: str | None) -> str:
        """Validate that `split` exists on the spec; return a concrete split."""

        if not spec.splits:
            return split or spec.default_split or "default"
        if split is None:
            if spec.default_split:
                return spec.default_split
            raise ValueError(f"split must be provided for dataset {spec.id}")
        if split not in spec.splits:
            raise ValueError(f"Unknown split '{split}' for dataset {spec.id}")
        return split

    @staticmethod
    def normalise_seed(seed: Any, *, cardinality: int | None = None) -> int:
        """Normalise arbitrary seed input into a bounded non-negative integer."""

        try:
            value = int(seed)
        except Exception as exc:  # pragma: no cover - defensive
            raise ValueError(f"Seed must be convertible to int (got {seed!r})") from exc
        if value < 0:
            value = abs(value)
        if cardinality and cardinality > 0:
            value = value % cardinality
        return value

File: synth_ai_core/src/localapi/datasets.rs
use crate::errors::CoreError;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskDatasetSpec {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub version: Option<String>,
    #[serde(default)]
    pub splits: Vec<String>,
    #[serde(default)]
    pub default_split: Option<String>,
    #[serde(default)]
    pub cardinality: Option<i64>,
    #[serde(default)]
    pub description: Option<String>,
}

impl TaskDatasetSpec {
    pub fn validate(&self) -> Result<(), CoreError> {
        if let Some(default_split) = &self.default_split {
            if !self.splits.is_empty() && !self.splits.contains(default_split) {
                return Err(CoreError::InvalidInput(
                    "default_split must be one of splits when provided".to_string(),
                ));
            }
        }
        Ok(())
    }

    pub fn merge_with(&self, other: &TaskDatasetSpec) -> TaskDatasetSpec {
        TaskDatasetSpec {
            id: if other.id.is_empty() {
                self.id.clone()
            } else {
                other.id.clone()
            },
            name: if other.name.is_empty() {
                self.name.clone()
            } else {
                other.name.clone()
            },
            version: other.version.clone().or_else(|| self.version.clone()),
            splits: if other.splits.is_empty() {
                self.splits.clone()
            } else {
                other.splits.clone()
            },
            default_split: other
                .default_split
                .clone()
                .or_else(|| self.default_split.clone()),
            cardinality: other.cardinality.or(self.cardinality),
            description: other
                .description
                .clone()
                .or_else(|| self.description.clone()),
        }
    }
}

pub fn ensure_split(spec: &TaskDatasetSpec, split: Option<&str>) -> Result<String, CoreError> {
    if spec.splits.is_empty() {
        return Ok(split
            .unwrap_or_else(|| spec.default_split.as_deref().unwrap_or("default"))
            .to_string());
    }
    match split {
        Some(value) => {
            if spec.splits.contains(&value.to_string()) {
                Ok(value.to_string())
            } else {
                Err(CoreError::InvalidInput(format!(
                    "Unknown split '{}' for dataset {}",
                    value, spec.id
                )))
            }
        }
        None => {
            if let Some(default_split) = &spec.default_split {
                Ok(default_split.clone())
            } else {
                Err(CoreError::InvalidInput(format!(
                    "split must be provided for dataset {}",
                    spec.id
                )))
            }
        }
    }
}

pub fn normalise_seed(seed: i64, cardinality: Option<i64>) -> i64 {
    let mut value = seed;
    if value < 0 {
        value = value.abs();
    }
    if let Some(card) = cardinality {
        if card > 0 {
            value = value % card;
        }
    }
    value
}
